lxd/cluster/upgrade.go

```diff
@@ -48,3 +54,76 @@ func NotifyUpgradeCompleted(state *state.State, cert *shared.CertInfo) error {
 		return nil
 	})
 }
+
+// KeepUpdated is a task that continuously monitor this node's version to see it
+// it's out of date with respect to other nodes. In the node is out of date,
+// and the LXD_CLUSTER_UPDATE environment variable is set, then the task
+// executes the executable that the variable is pointing at.
+func KeepUpdated(state *state.State) (task.Func, task.Schedule) {
+	f := func(ctx context.Context) {
+		ch := make(chan struct{})
+		go func() {
+			maybeUpdate(state)
+			close(ch)
+		}()
+		select {
+		case <-ctx.Done():
+		case <-ch:
+		}
+	}
+
+	schedule := task.Every(5 * time.Minute)
+
+	return f, schedule
+}
+
+// Check this node's version and possibly run LXD_CLUSTER_UPDATE.
+func maybeUpdate(state *state.State) {
+	shouldUpdate := false
+
+	enabled, err := Enabled(state.Node)
+	if err != nil {
+		logger.Errorf("Failed to check clustering is enabled: %v", err)
+		return
+	}
+	if !enabled {
+		return
+	}
+
+	err = state.Cluster.Transaction(func(tx *db.ClusterTx) error {
+		outdated, err := tx.NodeIsOutdated()
+		if err != nil {
+			return err
+		}
+		shouldUpdate = outdated
+		return nil
+	})
+
+	if err != nil {
+		// Just log the error and return.
+		logger.Errorf("Failed to check if this node is out-of-date: %v", err)
+		return
+	}
+
+	if !shouldUpdate {
+		logger.Debugf("Cluster node is up-to-date")
+		return
+	}
+
+	logger.Infof("Node is out-of-date with respect to other cluster nodes")
+
+	updateExecutable := os.Getenv("LXD_CLUSTER_UPDATE")
+	if updateExecutable == "" {
+		logger.Warn("No LXD_CLUSTER_UPDATE variable set, skipping auto-update")
+	}
+
+	logger.Infof("Triggering cluster update using: %s", updateExecutable)
+
+	output, err := shared.RunCommand(updateExecutable)
+	if err != nil {
+		logger.Errorf(err.Error())
```

Comment:
Should be wrapped with a useful error message or the log is going to be pretty confusing.

Reply:
Fixed.